catalyst/callbacks/metric.py

```diff
@@ -1,78 +1,300 @@
     ):
-        """@TODO: docs."""
+        """
+        Init MetricCallback
+
+        Args:
+            metric: metric to calculate in callback
+            input_key: keys of tensors that should be used as inputs in metric calculation
+            target_key: keys of tensors that should be used as targets in metric calculation
+        """
         super().__init__(order=CallbackOrder.metric, node=CallbackNode.all)
-        assert isinstance(metric, ICallbackBatchMetric)
         self.metric = metric
+        self._validate_metric(metric=metric)
+        self._metric_update_method = self.metric.update
+
+        is_value_input = isinstance(input_key, str)
+        is_value_targets = isinstance(target_key, str)
+
+        if is_value_input and is_value_targets:
+            self._get_inputs = self._get_value_inputs
+            self._update_metric = self._update_value_metric
+        elif not is_value_input and not is_value_targets:
+            self._get_inputs = self._get_key_value_inputs
+            self._update_metric = self._update_key_value_metric
+        else:
+            raise NotImplementedError()
+
         self.input_key = input_key
         self.target_key = target_key
-        self.log_on_batch = log_on_batch
+        self._keys = {
+            **self._convert_keys_to_kv(input_key),
+            **self._convert_keys_to_kv(target_key),
+        }
+
+    @staticmethod
+    def _validate_metric(metric: IMetric) -> None:
+        """
+        Check if given metric can be used in this callback
+
+        Args:
+            metric: metric to check
+        """
+        assert isinstance(metric, IMetric)
+
+    @staticmethod
+    def _convert_keys_to_kv(
+        keys: Union[str, Iterable[str], Dict[str, str]]
+    ) -> Dict[str, str]:
+        """
+        Convert keys to key-value format
+
+        Args:
+            keys: keys to convert
+
+        Returns:
+            dict of keys like {"a": "b"} where "a" is a field name of field in batch,
+                "b" is a name of the same data for metric
+        """
+        kv_keys = {}
+        if isinstance(keys, dict):
+            kv_keys.update(keys)
+        elif isinstance(keys, str):
+            kv_keys[keys] = keys
+        else:
+            for key in keys:
+                kv_keys[key] = key
+        return kv_keys
+
+    def _get_value_inputs(
+        self, runner: "Runner"
+    ) -> Tuple[torch.Tensor, torch.Tensor]:
+        """
+        Get data from batch in value input case
+
+        Args:
+            runner: current runner
+
+        Returns:
+            tuple of tensor of inputs and tensor of targets
+        """
+        inputs, targets = (
+            runner.batch[self.input_key],
+            runner.batch[self.target_key],
+        )
+        inputs, targets = (
+            runner.engine.sync_tensor(inputs),
+            runner.engine.sync_tensor(targets),
+        )
+        return inputs, targets
+
+    def _get_key_value_inputs(
```

Comment:
btw, the good thing about docs – you don't need to write them for the private methods.